kvbc/src/categorization/versioned_kv_category.cpp

```diff
@@ -0,0 +1,347 @@
+// Concord
+//
+// Copyright (c) 2020-2021 VMware, Inc. All Rights Reserved.
+//
+// This product is licensed to you under the Apache 2.0 license (the
+// "License").  You may not use this product except in compliance with the
+// Apache 2.0 License.
+//
+// This product may include a number of subcomponents with separate copyright
+// notices and license terms. Your use of these subcomponents is subject to the
+// terms and conditions of the subcomponent's license, as noted in the LICENSE
+// file.
+
+#include "categorization/versioned_kv_category.h"
+
+#include "assertUtils.hpp"
+#include "categorization/blockchain.h"
+#include "categorization/column_families.h"
+#include "categorization/details.h"
+
+#include <rocksdb/slice.h>
+#include <rocksdb/status.h>
+
+#include <utility>
+#include <variant>
+
+namespace concord::kvbc::categorization::detail {
+
+using namespace std::literals;
+
+template <typename Span>
+std::optional<std::string> value(const Span &ser) {
+  auto db_value = VersionedDbValue{};
+  deserialize(ser, db_value);
+  if (!std::holds_alternative<DbValue>(db_value.data)) {
+    return std::nullopt;
+  }
+  return std::get<DbValue>(std::move(db_value.data)).data;
+}
+
+template <typename Span>
+BlockId latestVersion(const Span &ser) {
+  auto latest_version = LatestKeyVersion{};
+  deserialize(ser, latest_version);
+  return latest_version.block_id;
+}
+
+VersionedKeyValueCategory::VersionedKeyValueCategory(const std::string &category_id,
+                                                     const std::shared_ptr<storage::rocksdb::NativeClient> &db)
+    : values_cf_{category_id + VERSIONED_KV_VALUES_CF_SUFFIX},
+      latest_ver_cf_{category_id + VERSIONED_KV_LATEST_VER_CF_SUFFIX},
+      db_{db} {
+  createColumnFamilyIfNotExisting(values_cf_, *db_);
+  createColumnFamilyIfNotExisting(latest_ver_cf_, *db_);
+}
+
+VersionedOutput VersionedKeyValueCategory::add(BlockId block_id,
+                                               VersionedInput &&in,
+                                               storage::rocksdb::NativeWriteBatch &batch) {
+  auto out = VersionedOutput{};
+  addDeletes(block_id, std::move(in.deletes), out, batch);
+  addUpdates(block_id, in.calculate_root_hash, std::move(in.kv), out, batch);
+  return out;
+}
+
+void VersionedKeyValueCategory::addDeletes(BlockId block_id,
+                                           std::vector<std::string> &&keys,
+                                           VersionedOutput &out,
+                                           storage::rocksdb::NativeWriteBatch &batch) {
+  const auto deleted = true;
+  const auto stale_on_update = false;
+  for (auto &&key : keys) {
+    auto versioned_key = VersionedRawKey{std::move(key), block_id};
+
+    updateLatestKeyVersion(versioned_key.value, block_id, batch);
```

Comment:
Currently, no. Instead, it will always be set to the latest version the key was updated or deleted at.

Reply:
I think I came to a conclusion here, that I'm about to implement for the BlockMerkle category. As discussed in a comment below, I think it is valuable to be able to know the version when the key was deleted. This is useful for conflict detection in pre-execution and by specifically preventing the [ABA problem](https://en.wikipedia.org/wiki/ABA_problem).

The problem I mainly had was that I didn't want to make the value more complex when stored in the index so that we can save space, and also time on serialization. Storing only a 64-bit int is ideal. However, I also don't think it's ideal to store whether or not the key is deleted in the value, as this requires two rocksdb lookups instead of 1 when you just want to do conflict detection on versions. 

Instead, the solution I'm going to implement is to just use the high bit as a tombstone flag. If it is 1, the key has been deleted at the version in the lower 63 bits. If 0, the key was written at that version. We can now only store 2^63 blocks, instead of 2^64, but I think we'll be ok with this :)

It's up to you if you want to use the same strategy for this category. Since you already have other flags in the value, and conflict detection isn't as important for these keys, maybe it doesn't matter. However, if you did use this strategy we could unify the interface for `getLatestVersion` such that it returns whether or not the key was deleted at that version. Something like:

```c++
struct LatestVersion {
    bool deleted;
    BlockId version;
};
```

Reply:
I made this change in https://github.com/vmware/concord-bft/pull/1120. Feel free to merge and use it if you want.

I'll follow up with the rest of the delete code tomorrow.

Reply:
I like that approach, I think I should do the same for the versioned category too as, AFAIK, it can also save state that is subject to conflict detection. Thanks!